src/runner_metrics.py

```diff
@@ -0,0 +1,198 @@
+#  Copyright 2023 Canonical Ltd.
+#  See LICENSE file for licensing details.
+
+"""Classes and function to extract the metrics from a shared filesystem."""
+
+import json
+import logging
+from json import JSONDecodeError
+from typing import Optional
+
+from pydantic import BaseModel, NonNegativeFloat, ValidationError
+
+import metrics
+import shared_fs
+from errors import CorruptMetricDataError, MetricFileSizeTooLargeError
+
+logger = logging.getLogger(__name__)
+
+FILE_SIZE_BYTES_LIMIT = 1024
+PRE_JOB_METRICS_FILE_NAME = "pre-job-metrics.json"
+POST_JOB_METRICS_FILE_NAME = "post-job-metrics.json"
+RUNNER_INSTALLED_TS_FILE_NAME = "runner-installed.timestamp"
+
+
+class PreJobMetrics(BaseModel):
+    """Metrics for the pre-job phase of a runner.
+
+    Args:
+        timestamp: The UNIX time stamp of the time at which the event was originally issued.
+        workflow: The workflow name.
+        workflow_run_id: The workflow run id.
+        repository: The repository name.
+        event: The github event.
+    """
+
+    timestamp: NonNegativeFloat
+    workflow: str
+    workflow_run_id: str
+    repository: str
+    event: str
+
+
+class PostJobMetrics(BaseModel):
+    """Metrics for the post-job phase of a runner.
+
+    Args:
+        timestamp: The UNIX time stamp of the time at which the event was originally issued.
+        status: The status of the job.
+    """
+
+    timestamp: NonNegativeFloat
+    status: str
+
+
+class RunnerMetrics(BaseModel):
+    """Metrics for a runner.
+
+    Args:
+        installed_timestamp: The UNIX time stamp of the time at which the runner was installed.
+        pre_job: The metrics for the pre-job phase.
+        post_job: The metrics for the post-job phase.
+    """
+
+    installed_timestamp: NonNegativeFloat
+    pre_job: PreJobMetrics
+    post_job: Optional[PostJobMetrics]
+
+
+def _inspect_file_sizes(fs: shared_fs.SharedFilesystem) -> None:
+    """Inspect the file sizes of the shared filesystem.
+
+    Args:
+        fs: The path to the shared filesystem for a specific runner.
+
+    Raises:
+        FileSizeTooLargeError: If one of the files is too large.
+    """
+    files = [
+        fs.path.joinpath(PRE_JOB_METRICS_FILE_NAME),
+        fs.path.joinpath(POST_JOB_METRICS_FILE_NAME),
+        fs.path.joinpath(RUNNER_INSTALLED_TS_FILE_NAME),
+    ]
+    for file in files:
+        if file.exists() and file.stat().st_size > FILE_SIZE_BYTES_LIMIT:
```

Comment:
I believe it was not mentioned in the spec but do you think if an invalid file is found, it should be deleted?

Reply:
Thanks for the suggestion. I think we should not remove the files that are too large, as we may want to inspect those files, as it could also be a programming error on our part (since we are writing to the file).

Reply:
I see. I was just concerned that they might be left behind and somehow create a problem. Maybe compress and move somewhere else.

Reply:
ps.: This can be addressed in a follow-up PR.

Reply:
Ah, I see. I think this should almost never happen, only if we really come across a malicious runner. 
The filesystem size is currently 1M, so it is unlikely to fill up the disk. I assume the operator would take immediate action if this happened, as the charm would go into error state.
@jdkandersson @yhaliaw What do you think?

Reply:
@amandahla So, after talking to Greg, we decided to go your way (move to quarantine). ;-)